### PR TITLE
feat: allow T.UnionOf to work with read only arrays

### PR DESCRIPTION
**Pull request by pengx17**
## 🎁 Pull Request

* [X] Used a clear / meaningful title for this pull request
* [X] Tested the changes in your own code (on your projects)
* [X] Added / Edited tests to reflect changes (`tst` folder)
* [X] Have read the **Contributing** part of the **Readme**
* [X] Passed `npm test`

#### Why have you made changes?

1. It is common to use `as const` to declare a tuple like the following:

```ts
const Foos = ['a', 'b'] as const;
type Foo = T.UnionOf<typeof Foos> // previously this will be an error
```

2. O.UnNest
Some times I want to get all possible types of the values of an object (used as a dict):

```ts
type Country = 'US' | 'UK' | 'CN'
type CountryCityMap = { US: ['NY', 'LA'], UK: ['London', 'Nottingham'], CN: ['Beijing', 'Shanghai'] };

// or O.UnNest<CountryCityMap>
// City = "NY" | "LA" | "London" | "Nottingham" | "Beijing" | "Shanghai"
type City = O.UnNest<CountryCityMap, Country> 
```

#### What changes have you made?
* changed this to achieve this
* changed that to achieve this
* ...

#### What tests have you updated?
* tested this in `tst/...`
* tested that in `tst/...`
* ...

#### Is there any breaking changes?

* [ ] Yes, I changed the public API & documented it
* [ ] Yes, I changed existing tests
* [ ] No,  I added to the public API & documented it
* [X] No,  I added to the existing tests
* [ ] I don't know

### CODE REVIEW

File: src/Object/UnNest.ts [hunk 1]
```diff
@@ -0,0 +1,15 @@
...
+
+/** Transform an **`object`**'s values into an **union**
+ * @param O to transform
+ * @returns **`any`**
+ * @example
+ * ```ts
+ * ```
+ */
+export type UnNest<O, K extends Index = keyof O> = O extends {
+    [key in K]?: infer B;
```
Comment (pirix-gh) [inline src/Object/UnNest.ts]: Wow, did not know this was possible

**Comment by pirix-gh** [comment]
Hey @pengx17 thanks for your pull request! I'm not going to accept it, but I'm going to help you out:

* To achieve what your type does, use `Path` instead:
```ts
type Country = 'US' | 'UK' | 'CN'
type CountryCityMap = { US: ['NY', 'LA'], UK: ['London', 'Nottingham'], CN: ['Beijing', 'Shanghai'] };

// City = "NY" | "LA" | "London" | "Nottingham" | "Beijing" | "Shanghai"
type City = Path<CountryCityMap, [Country, any]>
```
* And I've just pushed a new version that accepts `readonly` arrays for all the types, so you don't have to worry about this anymore :)

(cc @tonivj5)

**Comment by pengx17** [comment]
Thanks @pirix-gh. I saw you refactored all tuple types which is a great effort! Your repo on ts really inspires me a lot and thanks for maintaining it :D

**Comment by pirix-gh** [comment]
You're welcome @pengx17, thanks!

**Pull request closed without merging**